network-libp2p/src/connection_pool/behaviour.rs

```diff
@@ -115,6 +122,29 @@ struct ConnectionState<T> {
     required_services: Services,
 }
 
+enum ConnectionStateAction {
+    UnbanPeer(PeerId),
+}
+
+impl Stream for ConnectionState<PeerId> {
+    type Item = ConnectionStateAction;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        // Handle the first banned peer that can be unbanned
+        if let Some((peer_id, due_to)) = self.next_unban_peer {
+            let mut sleep_until_pin = pin!(sleep_until(due_to));
+
+            // Wait until the ban period has expired
+            if sleep_until_pin.poll_unpin(cx).is_ready() {
+                self.next_unban_peer = self.subsequent_unban_deadlines.pop_front();
+                return Poll::Ready(Some(ConnectionStateAction::UnbanPeer(peer_id)));
```

Comment:
I think it'd be better if we'd call `unmark_banned` directly **instead** of implementing a `Stream` here and instead of calling `unban_connection` on the parent struct, because
a) it's keeping the struct itself consistent (things in `banned` have corresponding deadlines.
b) we'll consistently unban the correct IP address after banning it, even if it might have changed in the meantime.